### PR TITLE
[3.3.5] Handle "CMSG_SET_SAVED_INSTANCE_EXTEND"

### PR DESCRIPTION
**Pull request by ccrs**
Based on some random work I found

### CODE REVIEW

File: src/server/game/Entities/Player/Player.h [hunk 1]
```diff
@@ -839,14 +839,23 @@ enum PlayerDelayedOperations
 // Maximum money amount : 2^31 - 1
 extern uint32 const MAX_MONEY_AMOUNT;
 
+#define INSTANCE_EXTEND_EXTEND      1
+#define INSTANCE_EXTEND_LOCK        2
```
Comment (DDuarte) [inline src/server/game/Entities/Player/Player.h]: I'd make this an enum class

```cpp
// def
enum class InstanceExtend
{
     Extend = 1,
     Lock     = 2
}

// usage:
bool IsLock() const { return (extend & InstanceExtend::Lock) != 0; }
```

File: sql/updates/world/2015_XX_XX_XX_world.sql [hunk 2]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE `character_instance` ADD COLUMN `extend` TINYINT(1) UNSIGNED NOT NULL DEFAULT '0';
+ALTER TABLE `instance` ADD COLUMN `lastresettime` int(10) unsigned NOT NULL DEFAULT '0';
```
Comment (tkrokli) [inline sql/updates/world/2015_XX_XX_XX_world.sql]: Why not use Capital column name `Extend` or UpperCamelCase in the compound name `LastResetTime`?

**Comment by ccrs** [comment]
I cant find any more errors, removing WIP from title.

**Comment by tkrokli** [comment]
Just remember to squash your commits so it is ready for merge. :+1:

**Comment by ccrs** [comment]
I'll squash when I receive more feedback, dont worry about it :)

**Comment by Nyeriah** [comment]
@Aokromes it's a tinyint on a table that holds temporary data (eventually if the values are 0 the rows will just get deleted at some point) so just adding that column won't hurt us that much

This looks fine as-is, just needs some more feedback I guess

**Comment by Aokromes** [comment]
I know it's a tinyint, but i don't see the logic of adding columns to tables for temporaly rarely used features :P

**Comment by NorthStrider** [comment]
@Aokromes deny it as much as you want but it is fact that this is a instance save feature regardless if it is used rarely or often.

**Comment by Aokromes** [comment]
NorthStrider, that increases even more the reason to be separate table :)

**Comment by P-Kito** [comment]
Did you even learn normalizing db structures :+1: :-P

Anyways, nice work so far!

**Comment by Aokromes** [comment]
@P-Kito my future plans: :p

http://pastebin.com/b8F7CumV

**Comment by joschiwald** [comment]
we could change the `permanent` column to a `flag` column

**Comment by Nyeriah** [comment]
Any feedback on joschiwald's suggestion? Myself I guess it could work out

**Comment by ccrs** [comment]
humm I think that work should be done by devs, not by me. It's not "entirely" related to this PR, in my opinion.

**Comment by mik1893** [comment]
Squash squash squash!!!!

**Comment by ccrs** [comment]
Done, but still needs tests

File: src/server/game/Instances/InstanceSaveMgr.h [hunk 3]
```diff
@@ -74,6 +74,9 @@ class InstanceSave
         time_t GetResetTime() const { return m_resetTime; }
         void SetResetTime(time_t resetTime) { m_resetTime = resetTime; }
         time_t GetResetTimeForDB();
+        time_t GetDefaultResetTime();
+        time_t GetNextResetTime();
+        bool IsExpired() { return time(NULL) > m_resetTime; }
```
Comment (DDuarte) [inline src/server/game/Instances/InstanceSaveMgr.h]: Missing a const in the method

File: sql/updates/characters/2015_XX_XX_XX_characters.sql [hunk 4]
```diff
@@ -0,0 +1 @@
+ALTER TABLE `character_instance` ADD COLUMN `extend` TINYINT(1) UNSIGNED NOT NULL DEFAULT '0';
```
Comment (DDuarte) [inline sql/updates/characters/2015_XX_XX_XX_characters.sql]: This change should be added to `character_database.sql` as well.

File: src/server/game/Instances/InstanceSaveMgr.h (hunk 3 above)
Comment (DDuarte) [inline src/server/game/Instances/InstanceSaveMgr.h]: GetNextResetTime, GetDefaultResetTime and GetResetTimeForDB should also be const

**Comment by jackpoz** [comment]
any follow up to latest comments ?

**Comment by Shauren** [comment]
I don't see one thing here, critical to extending expired saves
Where do you handle skipping deletion of instance state during weekly reset? There is no point in handling the opcode if you don't have data from `instance`, `creature_respawn` and `gameobject_respawn` for the saves? You would essentially be walking into a fresh instance with only the 'extend' flag set

**Comment by jackpoz** [comment]
@ccrs any followup to https://github.com/TrinityCore/TrinityCore/pull/14990#issuecomment-126684742 ?

**Comment by Nyeriah** [comment]
> @ccrs any followup to #14990 (comment) and all previous comments ?

```
This branch has conflicts that must be resolved
 Use the command line to resolve conflicts before continuing.
```

**Comment by ccrs** [comment]
been busy lately, but id like to finish this. Most important part is @Shauren comment.

I think it can be solved by bosstate: if boss is done, then dont respawn (linked_respawn will prevent trash respawn)

any thoughts?

**Comment by jackpoz** [comment]
Please ask a TrinityCore member to reopen this Pull Request once it's complete. There is no need to open a new Pull Request with the same changes.

**Pull request closed without merging**